### PR TITLE
Feat/ethereum get asset transfers

### PR DESCRIPTION
**Pull request by brucexc**
- add getAssetTransfers func from datasource etl to replace alchemy

### CODE REVIEW

File: common/datasource/pregod_etl/ethereum/client.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	FromAddress string `json:"fromAddress,omitempty"`
+}
+
+type GetAssetTransfersResult struct {
+	Transfers []Transfer
+}
+
+type Transfer struct {
+	BlockNum int64   `gorm:"column:block_number"`
+	Hash     []byte  `gorm:"column:hash"`
```
Comment (kallydev) [inline common/datasource/pregod_etl/ethereum/client.go]: `common.Hash`

File: common/datasource/pregod_etl/ethereum/client.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	From     []byte  `gorm:"column:from_address"`
```
Comment (kallydev) [inline common/datasource/pregod_etl/ethereum/client.go]: `common.Address`

File: common/datasource/pregod_etl/ethereum/client.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	Value    float64 `gorm:"column:value"`
+	Category string  `gorm:"column:category"`
+}
+
+func GetAssetTransfers(ctx context.Context, parameter GetAssetTransfersParameter) (*GetAssetTransfersResult, error) {
+	result := GetAssetTransfersResult{}
+	fromAddress := "\\" + parameter.FromAddress[1:]
+	blockNum := parameter.FromBlock
+
+	err := database.EthDb().
```
Comment (kallydev) [inline common/datasource/pregod_etl/ethereum/client.go:33]: Only transactions from `message.Address` will be fetched here, which will cause the address whitelist to become useless. 🤔

File: service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+	// Get all block and transaction data
+	for _, transaction := range transactionMap {
+		internalTransaction := transaction
+
+		if internalTransaction.BlockNumber < message.BlockNumber {
+			continue
+		}
+
+		if internalTransaction.AddressFrom != "" && !strings.EqualFold(internalTransaction.AddressFrom, message.Address) && !allowlist.AllowList.Contains(internalTransaction.AddressFrom) {
```
Comment (kallydev) [inline service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go]: This is the code associated with the whitelist.

File: service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	return internalTransactions, nil
+}
+
+func New() datasource.Datasource {
+	return &Datasource{}
+}
+
+func AppendHexPrefix(hexaString string) string {
```
Comment (kallydev) [inline service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go]: [https://github.com/ethereum/go-ethereum/tree/master/common/hexutil](https://github.com/ethereum/go-ethereum/tree/master/common/hexutil).

**Review by kallydev — CHANGES_REQUESTED** [review]

File: common/datasource/pregod_etl/ethereum/client.go (hunk 2 above)
Comment (brucexc) [reply common/datasource/pregod_etl/ethereum/client.go]: if set type to_address as common.Address , error " can't scan []byte of len 0 into Address, want 20" will be existed  when to_address is null

File: common/datasource/pregod_etl/ethereum/client.go (hunk 3 above)
Comment (brucexc) [reply common/datasource/pregod_etl/ethereum/client.go:33]: etl don't need to deal with this whitelist logic because neither poap nor lens is on ethereum

Comment (polebug) [reply common/datasource/pregod_etl/ethereum/client.go:33]: I think database query should use pagination?

File: service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/naturalselectionlabs/pregod/common/protocol"
+	"github.com/naturalselectionlabs/pregod/common/utils/loggerx"
+	"github.com/naturalselectionlabs/pregod/common/utils/opentelemetry"
+	"github.com/naturalselectionlabs/pregod/service/indexer/internal/datasource"
+	"go.opentelemetry.io/otel"
+	"go.uber.org/zap"
+)
+
+const (
+	Source = "ethereum-etl"
```
Comment (polebug) [inline service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go]: maybe use `protocol.SourcePregodETL`
i think lens alse should use it

File: common/datasource/pregod_etl/ethereum/client.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package ethereum
```
Comment (polebug) [inline common/datasource/pregod_etl/ethereum/client.go:1]: I don't think there is a need to open a directory for ethereum.
`GetAssetTransfers` is common

File: service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	// Get all block and transaction data
+	for _, transaction := range transactionMap {
+		internalTransaction := transaction
+
+		if internalTransaction.BlockNumber < message.BlockNumber && !strings.EqualFold(internalTransaction.AddressFrom, message.Address) {
+			continue
+		}
+
```
Comment (kallydev) [inline service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go:59]: Why deleted allowlist? 👀

**Review by kallydev — CHANGES_REQUESTED** [review]

File: service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go (hunk 8 above)
Comment (brucexc) [reply service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go:59]: i don't think ethereum has poap address in allowlist

File: common/datasource/pregod_etl/ethereum/client.go (hunk 7 above)
Comment (brucexc) [reply common/datasource/pregod_etl/ethereum/client.go:1]: But the current situation is that different network is connected to different ETL source data，which leads to that  GetAssetTransfers is uncommon 🤔

File: service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go (hunk 6 above)
Comment (brucexc) [reply service/indexer/internal/datasource/pregod_etl/ethereum/datasource.go]: great

**Pull request merged**